diff --git a/globals.c b/globals.c
index a786a4b9..1b2b9ad0 100644
--- a/globals.c
+++ b/globals.c
@@ -118,6 +118,13 @@ static XML_THREAD_LOCAL xmlGlobalState globalState;
     defined(__GLIBC__) && \
     __GLIBC__ * 100 + __GLIBC_MINOR__ < 234
 
+#pragma weak pthread_getspecific
+#pragma weak pthread_setspecific
+#pragma weak pthread_key_create
+#pragma weak pthread_key_delete
+#pragma weak pthread_equal
+#pragma weak pthread_self
+
 #define XML_PTHREAD_WEAK
 
 static int libxml_is_threaded = -1;
@@ -566,7 +573,9 @@ void xmlInitGlobalsInternal(void) {
             (pthread_getspecific != NULL) &&
             (pthread_setspecific != NULL) &&
             (pthread_key_create != NULL) &&
-            (pthread_key_delete != NULL);
+            (pthread_key_delete != NULL) &&
+            (pthread_equal != NULL) &&
+            (pthread_self != NULL);
     if (libxml_is_threaded == 0)
         return;
 #endif /* XML_PTHREAD_WEAK */
